### ISSUE TITLE
Authentication is not working in TypeScript Skill

### ISSUE THREAD

**Issue by tomSauret847** [microsoft/botframework-solutions#3573]
#### What project is affected?
TypeScript Virtual assistant and Skill
#### What language is this in?
TypeScript
#### What happens?
The following error is thrown when attempting authentication

Error: DialogContext.beginDialog(): A dialog with an id of 'AuthPrompt' wasn't found.
    at WaterfallStepContext.<anonymous> (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:123:23)
    at Generator.next (<anonymous>)
    at D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:3:12)
    at WaterfallStepContext.beginDialog (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:119:16)
    at MultiProviderAuthDialog.firstStep (D:\home\site\wwwroot\node_modules\bot-solutions\lib\authentication\multiProviderAuthDialog.js:75:34)
    at WaterfallDialog.<anonymous> (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\waterfallDialog.js:192:48)
    at Generator.next (<anonymous>)
    at D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\waterfallDialog.js:7:71

#### What are the steps to reproduce this issue?
Deploy virtual Assistant and skill. Enable the MultiProviderAuthDialog in the skill
Use the authentication dialog to authenticate with Azure Active Directory v2
#### What were you expecting to happen?
Receive login card to sign in
#### Can you share any logs, error output, etc.?
Error: DialogContext.beginDialog(): A dialog with an id of 'AuthPrompt' wasn't found.
    at WaterfallStepContext.<anonymous> (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:123:23)
    at Generator.next (<anonymous>)
    at D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:7:71
    at new Promise (<anonymous>)
    at __awaiter (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:3:12)
    at WaterfallStepContext.beginDialog (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\dialogContext.js:119:16)
    at MultiProviderAuthDialog.firstStep (D:\home\site\wwwroot\node_modules\bot-solutions\lib\authentication\multiProviderAuthDialog.js:75:34)
    at WaterfallDialog.<anonymous> (D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\waterfallDialog.js:192:48)
    at Generator.next (<anonymous>)
    at D:\home\site\wwwroot\node_modules\botbuilder-dialogs\lib\waterfallDialog.js:7:71

**Comment by Batta32** [comment]
Thanks @tomSauret847 for reporting this issue! As soon as we have any update, we will back to you 😊.

**Comment by Batta32** [comment]
Hi @tomSauret847 - sorry about the delay. We have successfully **reproduced** the issue using the TypeScript Skill following these steps:

1. Uncomment the lines in [skillDialogBase](https://github.com/microsoft/botframework-solutions/blob/master/templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts#L48) marked for authentication
1. Pass an `OAuthPromptSettings` array to the `MultiProviderAuthDialog` constructor
1. Add `appsettings.oauthConnections` to the [properties](https://github.com/microsoft/botframework-solutions/blob/master/sdk/typescript/libraries/bot-solutions/src/botSettings.ts#L37) of `botSettings`
1. Add an `AADv2` connection
1. Set up an `AADv2` connection to the registration in Azure

As soon as we have any update, we will back to you 😊.

![image](https://user-images.githubusercontent.com/11904023/89198324-6d55e800-d583-11ea-8a69-211c0b2c0856.png)

**Comment by Batta32** [comment]
Hi @tomSauret847 - we have **fixed** this issue in this [commit](https://github.com/microsoft/botframework-solutions/pull/3583/commits/ac27880bda35df549935adb965453c85cc18883f).

This was actually a problem in the `bot-solutions` library, and we have scheduled the fix to be released with the version [1.0](https://github.com/microsoft/botframework-solutions/pull/3583) of the library.

In the meantime, if you want a way to test the fix for yourself, you can follow these **steps**:
1. Checkout this [branch](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill), with the fix applied for version 0.8 of the library.
1. Go to the [bot-solutions](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill/sdk/typescript/libraries/bot-solutions) folder
1. Execute `npm install` to install its dependencies.
1. Execute `npm run build` to build the solution.
1. Execute `npm pack` to generate a `.tgz` file. It should create a file at the same location, named bot-solutions-version.tgz
1. Point at the generated tgz in your Skill bot, replacing the bot-solutions reference in the bot's package.json from `"bot-solutions": "^1.0.0"` to `"bot-solutions": "<PATH_TO_TGZ>"`
1. Validate the authentication using the Skill

![image](https://user-images.githubusercontent.com/11904023/89538757-573d6700-d7d1-11ea-8570-e592aa450877.png)

We will be attentive to your replies 😊.

**Comment by tomSauret847** [comment]
Thank you for looking into this @Batta32 I am still receiving an error on this process. The error is 
Error: Could not find Connection Setting with name {connection name}

I am setting it up the connection as outlined [here ](https://microsoft.github.io/botframework-solutions/overview/whats-new/1.0/enable-sso-with-skills-using-oauthcredentials-setting/) for setting up the SSO for a skill.

Do we still need to have a authentication connection in each skill with the typescript skills?
Or can we use the single SSO like it is outlined for the C# skills? 

If I add the connection on the skill I will get the sign in prompt as you did but I will not receive the token back after completing the log in. If you could clarify if we are needing to setup a single connection in the VA or if we need to setup connections in each of the skills that require authentication.

**Comment by Batta32** [comment]
Thanks @tomSauret847 for your reply. As soon as we have any update, we will back to you 😊!

**Comment by Batta32** [comment]
Hi @tomSauret847 - we have successfully **reproduced** the issue. It's because the `name` property in `oauthConnections` of the `appsettings.json` differs to the `connectionName` property in the `OAuthPromptSettings` you are using.

We will continue reviewing this and analyzing the questions you have mentioned above. As soon as we have any update, we will back to you 😊.

The **steps** we followed to reproduce the issue were:
1. Follow the steps of this [comment](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-668085672)
6. After this, follow the steps of this [comment](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-669935733)
12. Check that the `name` property in `oauthConnections` of the `appsettings.json` differs to the `connectionName` property in the `OAuthPromptSettings` you are using
13. Start the Skill and type "run sample dialog", and see the error

_When connectionName and name are different, the issue is reproduced_
![image](https://user-images.githubusercontent.com/11904023/89816294-ae6a7100-db1c-11ea-95d2-0793053e5319.png)

_Issue reproduced_
![image](https://user-images.githubusercontent.com/11904023/89816302-b4605200-db1c-11ea-8425-d8168f15fc18.png)

**Comment by Batta32** [comment]
Hi @tomSauret847

> Do we still need to have a authentication connection in each skill with the typescript skills?

You need to have an authentication connection in each TypeScript Skill until the version [1.0](https://github.com/microsoft/botframework-solutions/pull/3585) of TypeScript Skill is released.

> Or can we use the single SSO like it is outlined for the C# skills?

SSO for TypeScript can't be used as it is outlined for the C# Skills since the version 1.0 of TypeScript is not released yet.

Let us know if these helps you 😊.

**Comment by Batta32** [comment]
@tomSauret847 - the error `Error: Could not find Connection Setting with name {connection name}` is because the `name` property in `oauthConnections` of the `appsettings` differs to the `connectionName` property in the `OAuthPromptSettings` you are using.

To solve this, you have to check that both properties are **equal**.

For more information, you can check the [comment](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-671512787) of above.

**Comment by tomSauret847** [comment]
Thank you @Batta32 for looking into this. I can confirm that if I create a connection property on the skill I can get the login prompt. Your comment has brought up another question. We are getting ready to move into testing with our new VA and after that into production. The question I have,

Is this fix for the authentication going to be published to NPM before the 1.0 release?

For our production environment we will need published NPM packages so we would need this fix to be published before we can move this VA into production. I will be attentive to your answer.

**Comment by Batta32** [comment]
@tomSauret847 - Thanks for your reply! We are adding these fixes in the PRs of TypeScript version 1.0:
- PR #3583: `[TypeScript][Bot-Solutions] Implement changes in Bot-Solutions to 1.0 release`
- PR #3584: `[TypeScript][Virtual Assistant] Implement changes in Virtual Assistant to 1.0 release`
- PR #3585: `[TypeScript][Skill] Implement changes in Skill to 1.0 release`

We will inform you as soon as we have any update on this. Meanwhile, we are working on validating the Authentication process in TypeScript.

**Comment by Batta32** [comment]
@tomSauret847 - we identified another issue that the [Skill State](https://github.com/microsoft/botframework-solutions/blob/master/templates/typescript/samples/sample-skill/src/dialogs/skillDialogBase.ts#L78) variable was _undefined_ during the Authentication process.

We took as a guide the PR #3561 which sets a default value for the Skill State accessor that we incorporated in this [commit](https://github.com/microsoft/botframework-solutions/pull/3585/files/31f31376bf77399f242ee76ee826646ea7b8f38d..71ef82354e554bc1f3494df5b2921ebb397a76cb).
After applying these changes, the authentication process correctly works. This issue will be fixed as soon as the TypeScript v1.0 is merged.

This is our environment using this [branch](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill): `feature/southworks/test-fix-auth-skill`:
- [TypeScript Skill Sample](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill/templates/typescript/samples/sample-skill)

In the meantime, if you want a way to test the fix for yourself, you can follow these steps:

1. Go to the [bot-solutions](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill/sdk/typescript/libraries/bot-solutions) folder
2. Execute `npm install` to install its dependencies.
3. Execute `npm run build` to build the solution.
4. Execute `npm pack` to generate a .tgz file. It should create a file at the same location, named bot-solutions-version.tgz
5. Go to the [Skill Sample](https://github.com/southworks/botframework-solutions/tree/feature/southworks/test-fix-auth-skill/templates/typescript/samples/sample-skill) folder
6. Point at the generated tgz in the Skill Sample bot, replacing the bot-solutions reference in the bot's package.json from `"bot-solutions": "^1.0.0"` to `"bot-solutions": "<PATH_TO_TGZ>"`
7. Execute `npm install` to install its dependencies.
8. Execute `npm run build` to build the solution.
9. Validate the authentication process

We're attentive to your answer. 😊

_SkillState variable getting an undefined value_
![image](https://user-images.githubusercontent.com/11904023/90031529-b862b000-dc93-11ea-86ae-f5b465a1a7ac.png)

_Authentication process correctly working_
![image](https://user-images.githubusercontent.com/11904023/90031520-b567bf80-dc93-11ea-8697-582b29f41f40.png)

**Comment by tomSauret847** [comment]
Thank you again @Batta32! I did use the [1.0](https://github.com/microsoft/botframework-solutions/pull/3585) package for the testing. I still need to keep a connection setting in the skill, but I am able to get the authentication to work. I was able to see where the VA is receiving the token response and forwarding that to the skill to process.

**Comment by tomSauret847** [comment]
@Batta32 We have setup authentication in 2 of our skills and it is working good in the Teams channel. The direct line channel is not getting past the authentication prompt. The skill is not receiving the token back after successful login. This issue is only present in our direct line channel. We are using the sample direct line client that was provided and enabled  the enhanced authentication options so we do not enter the "Magic code". It worked before adding the enhanced authentication but after adding this feature the skill is no longer receiving the token response back. If we cancel out of the skill and go back into that dialog the token will be present and we can complete the dialog. Let me know if you have any thoughts on what might be causing this.

**Comment by Batta32** [comment]
Thanks @tomSauret847, we will be reviewing that scenario and we will back to you as soon as we can!

Just to start to review the scenario:
1. What do you mean with "sample direct line client"?
1. Which client are you using?
1. What do you mean with "Magic Code"?

**Comment by tomSauret847** [comment]
Thank you again @Batta32 for looking into this. We are using the direct line sample from the botbuilder-solutions repo located [here](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients) 
The "magic code" is explained some in this stack overflow question [here](https://stackoverflow.com/questions/57053009/magic-code-while-oauth-using-botframework-any-way-around)
It is just the code the bot user is required to copy and paste back into the conversation after successfully authenticating. By implementing the  enhanced authentication it removes this step for the bot user so they only need to authenticate and the conversation moves forward as the token is passed back to the bot behind the scenes.

**Comment by Batta32** [comment]
Thanks a lot @tomSauret847, we will be reviewing this and we will let you know any update here 😊!

**Comment by matiasroldan6** [comment]
Hi @tomSauret847, we **could not reproduce** this issue as we have the authentication scenario correctly working.

We came up with some **questions** for you:

1. Have you checked if this issue happens talking directly to the Skill, without connecting to a Virtual Assistant?
2. If you are going through a V.A. to connect to the Skill, are you enabling the Directline channel in the V.A. too? What else have you set in it, Azure AD v2 auth, enhanced authentication?
3. What do you mean with "If we cancel out of the skill and go back into that dialog the token will be present and we can complete the dialog"?

This is the setup we've been using:

- A TypeScript Skill created with the generator-bot-virtualassistant@1.0.1
- Azure AD v2 type authentication set-up on that Skill
- Directline channel enabled with enhanced authentication on the Skill
- The Directline [web sample](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web) from the repository set to a specific user id ("dl_xxxx")
- The appsettings.json at the Directline client set similarly to this:

```JSON
{
    "Logging": {
        "LogLevel": {
            "Default": "Warning"
        }
    },
    "AllowedHosts": "*",
    "BotName": "skillbot-name",
    "DirectLineSecret": "XXXXXXXXXXXXXXXXXXXXXXXXXXXX",
    "EnableDirectLineEnhancedAuthentication": true,
    "SpeechServiceRegionIdentifier": "",
    "SpeechServiceSubscriptionKey": ""
}
```

The steps we followed are:

1. Contacting the Skill through the Directline client
2. Invoking the login dialog ("run sample dialog")
3. Clicking on the login button (no need for the "magic code")
4. We successfully login and the bot greets us
5. From that time on, every time we repeat steps 1 and 2 we are automatically logged in and greeted by the bot

We will be attentive to your answer.

**Comment by tomSauret847** [comment]
Thank you @matiasroldan6 for looking into this. To answer your questions
1. We are unable to contact the skill directly as we have it inside a VNET so it cannot be reached by the channels
2. We are reaching the skill through the VA and have the direct line channel configured on just the VA. We are using Azure AD V2 auth in the connection properties on the Skill and VA. We have the enhanced Authentication set on the VA
3. if we complete the sign in when presented with the sign in card. The bot will hang and not move forward. If we cancel then go back into a skill that has authentication it has the token and bypasses the signin prompt and allows us to continue. We have 2 skills with authentication and if we authenticate in one we are authenticated in the second as well.
here is a screen shot of our configuration for the direct line channl on the VA
![image](https://user-images.githubusercontent.com/61023857/92025604-6c9ca700-ed25-11ea-8e5b-47a2b36f70f7.png)

We do have the user ID set to "dl_xxxx" in the direct line client
We have the Azure AD v2 authentication in the connection on the Skill and VA
We have the appsettings.json of the direct line client set the same as you have above
We are unable to use the direct line channel on the skill and only use it on the VA
We are able to use Authentication in the Teams channel with the only issue is the dialog messages are out of order. It is failing to send the first stepContext.context.sendActivity('message') until a minute later in the conversion. This only happens on dialogs that have authentication include in them and in the Teams channel only.
We are using the bot-solutions 1.0 that is located [here](https://github.com/microsoft/botframework-solutions/pull/3583)

Please let me know if you need anymore information on the skill and VA

**Comment by matiasroldan6** [comment]
Hi @tomSauret847, we **could reproduce** an occurrence of this issue.
In our scenario, the problem were the **trusted origins addresses** set in the Directline channel. For instance, the Directline popup states that the trusted origins **need to be** _https_, or _http_ for _localhost_. This would explain Directline as the only channel that's failing for you.

To check in case this is also your problem, can you try to set as trusted origins:
Your bot address (either https://xxxxx.azurewebsites.net or https://xxxx.ngrok.io if you are debugging locally)
Your localhost if you need it, as http://localhost
![image](https://user-images.githubusercontent.com/20074735/92275065-c8e9fd00-eec4-11ea-995b-fab1ca70c743.png)


**The setting to reproduce this was:**
- A [1.0 Virtual Assistant](https://github.com/microsoft/botframework-solutions/pull/3584) with
   - [Bot-Solutions 1.0](https://github.com/microsoft/botframework-solutions/pull/3583) at the dependencies
   - Azure AD v2 authentication enabled
   - Directline channel enabled with enhanced authentication on
   - Trusted origins in the Directline channel set to https://localhost:xxxx (_xxxx_ being the Directline client address) and the bot address (https://xxxxx.azurewebsites.net or https://xxxx.ngrok.io for local connections). We did this emulating the settings from the picture you attached in your last message.
- A [1.0 Skill](https://github.com/microsoft/botframework-solutions/pull/3585) with
   - [Bot-Solutions 1.0](https://github.com/microsoft/botframework-solutions/pull/3583) at the dependencies
   - Azure AD v2 authentication enabled
   - Connected to the Virtual Assistant.
- The [Directline web client](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web) set with user id and enhanced authentication with the same configuration from our previous message.

**Our repro steps were:**
1. Connect the Directline sample to the Assistant.
2. Attempt to trigger authentication ("run sample dialog").
3. The Authentication dialog shows up but after clicking the button nothing happens, and the bot becomes unresponsive, even to further attempts to invoke authentication.
![image](https://user-images.githubusercontent.com/20074735/92274834-55e08680-eec4-11ea-9ac4-c10f85858f26.png)


4. Cancel the flow and then re-run steps 1 and 2, then we get automatically authenticated and we get greeted by the bot.

Please let us know if this is of any help. We'll be attentive to your responses.

**Comment by tomSauret847** [comment]
Thank you @matiasroldan6 for looking into this. The settings you show above are the same as we are using. We removed the localhost route and just have the bot address in the trusted origins for the direct line channel (https://xxxxx.azurewebsites.net). The skill is still unresponsive after the sign in prompt and if we cancel the conversation and invoke the skill again it will have the token and continue the dialog. We are still unable to get the skill to receive the token after the user signs in.

**Comment by Batta32** [comment]
Thanks for your reply @tomSauret847 - we will keep working on this trying to find the fix for this issue 😊.

We have **questions** in order to understand how you are using the `azurewebsites.net` endpoint taking into account the version 1.0 of bot-solutions (PR #3583):
1. Are you using the `Messaging endpoint` of your published Web App Bot (Virtual Assistant)?
1. How did you integrate the version 1.0 of bot-solutions in your published Virtual Assistant and Skill? We understand that you used the `.tgz` package of the library. Is this correct?
1. Is there any possibility to test using the localhost endpoint configuring it as [above](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-687317382) in the Direct Line Channel configuration?

**Comment by tomSauret847** [comment]
Thank you @Batta32 
1. We are using the messaging endpoint of the VA in the trusted host configuration. 
2. we are using the .tgz package that was cloned from the 1.0 release. We removed the node_modules folder and the package-lock.json files prior to moving over the the .tgz package. Then updated the package.json file and then ran the npm install
3. We have tested with the local host and received the same response. We are still unable to get the token to pass through the conversation, but when the dialog is canceled and started again the token is present.

**Comment by Batta32** [comment]
Thanks @tomSauret847, we will continue reviewing this issue with this new information that you provided! As soon as we have any update, we will back to you 😊.

**Comment by matiasroldan6** [comment]
Hi @tomSauret847, we successfully reproduced an occurrence of this issue.

One thing we noticed from your [previous responses](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-689510502) is that you're using your bot URL as a trusted origin.
 According to [this comment](https://github.com/microsoft/BotFramework-Services/issues/100#issuecomment-441957719), and [this documentation](https://blog.botframework.com/2018/09/25/enhanced-direct-line-authentication-features/), you should have the URL of your chat client.
![image](https://user-images.githubusercontent.com/20074735/94314918-f2dc8f80-ff57-11ea-9d90-1ce2779c7123.png)


**Our setup was:**
- [1.0 Bot-Solutions](https://github.com/microsoft/botframework-solutions/pull/3583)
- [1.0 Virtual Assistant](https://github.com/microsoft/botframework-solutions/pull/3584) deployed to Azure with:
   - Azure AD v2 authentication connection
   - Directline channel enabled with enhanced authentication
- [1.0 Skill](https://github.com/microsoft/botframework-solutions/pull/3585) deployed to Azure with:
   - Azure AD v2 authentication connection
- [Directline web client](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web) (C#) with Enhanced Authentication enabled

**Steps:**

1. Use the Directline web client to contact the Virtual Assistant
2. Invoke the Skill ("run sample dialog")
3. Be presented the authentication button by the Skill, click on it (nothing happens)
4. Repeating step 2 authenticates us automatically, without displaying the authentication button.

We will continue reviewing this issue as we confirmed and as you mentioned, that in Microsoft Teams and Emulator this is working.
![image](https://user-images.githubusercontent.com/20074735/94315126-464edd80-ff58-11ea-8dbd-b73c87e0e76b.png)

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, we **successfully** executed the authentication process using Enhanced Authentication.

We came up with some **questions**  so you can check in our side:

1.  Can you check if you have the published folders of the deployed bots accessing in Kudu (https://<WEB_APP_BOT_NAME>.scm.azurewebsites.net/ZipDeployUI) replacing `<WEB_APP_BOT_NAME>` with the name of your bot resource? Check if you have the `node_modules` , `lib` , and the rest of the folders.
2.  Can you check if the Messaging Endpoint of the bots is the production endpoint (azurewebsites.net)?
3.  Have you connected the Virtual Assistant to the Skills using the production endpoint?
4.  Can you check if you can access to the Skill Manifest of the published Skill? If not, add the changes of the PR [#3601](https://github.com/microsoft/botframework-solutions/pull/3601)

This is our **environment**:

-   1.0 TypeScript Virtual Assistant Sample (PR [#3584](https://github.com/microsoft/botframework-solutions/pull/3584))
-   1.0 TypeScript Skill Sample (PR [#3585](https://github.com/microsoft/botframework-solutions/pull/3585))
-   1.0 Bot-Solutions (PR [#3583](https://github.com/microsoft/botframework-solutions/pull/3583))
-   [Sample Direct Line Channel](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web)
-   Enhanced Authentication configured in the Virtual Assistant

We followed these **steps**:

1. Apply the changes of the PR [#3601](https://github.com/microsoft/botframework-solutions/pull/3601) in the Skill
2. Create a folder to store the Bot-Solutions 1.0 `.tgz` file on both samples
3. Point the bot-solutions reference to the local tgz, such as `"bot-solutions": ".//TGZ//bot-solutions-1.0.0.tgz"` of the bots
4. Deploy the bots using `deploy.ps1`
5. Connect both samples with Botskills using the remoteManifest
6. Configure the OAuth Connection Settings of both samples with an AADv2
7. Configure the oauthConnections in the appsettings of each sample
8. Publish both samples using the `publish.ps1` script
9. Activate Directline channel on the Virtual Assistant
10. Enable Enhanced Authentication on the Directline of the Virtual Assistant
11. Connect the Directline sample to the Virtual Assistant
      1.  **Note**: [Update this line](https://github.com/microsoft/botframework-solutions/blob/master/samples/csharp/clients/directline.web/Controllers/HomeController.cs#L35) to a static GUID if you wish to preserve the conversation
12. Finally, trigger the authentication using `run sample dialog`
13. After login in, the Skill will ask for a name
14. Check that the enhanced authentication is working correctly

Last but not least, we **confirmed** the same steps using C# bots and checked that is correctly working too.

We will be attentive to your reply 😊

_Configuration of the Enhanced Authentication_
![image](https://user-images.githubusercontent.com/39467613/94860998-37ab6f00-040d-11eb-9a0b-733d356e22de.png)

_Successful communication using Enhanced Authentication and TypeScript bots_
![auth fixed](https://user-images.githubusercontent.com/39467613/94861034-4560f480-040d-11eb-9892-811904c62945.gif)

_Successful communication using Enhanced Authentication and C# bots_
![image](https://user-images.githubusercontent.com/39467613/94861073-5447a700-040d-11eb-9b4f-eafbe91df87c.png)

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk. Sorry for the late response we needed to setup a testing instance to verify the fix. We have been able to get the authentication to work in the test instance and have narrowed down where the issue is.

For our instance of the VA /Skills we have them behind a app gateway with a firewall. It appears this is the reason the enhanced authentication is not working. When we implement this on a VA and skill outside of that configuration it is working properly. the app gateway places another host in the mix and we are thinking this is what is blocking the token.  So we have the following question,

Is there a way to get the enhanced authentication to work with a bot behind an App Gateway?

**Comment by Batta32** [comment]
Great @tomSauret847! We will do a research and some tests in order to review a way to get the enhanced authentication working with a bot behind an App Gateway.

As soon as we have any update for this, we will let you know here 😊.

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, we did a little research, and we came up with some **questions**.

1. Are you testing using the bots deployed in Azure or on promises' server?
2. Is your application gateway an [Azure Application Gateway](https://docs.microsoft.com/en-us/azure/application-gateway/overview)?
3. Are you using TypeScript or C# bots? As you [previously mentioned](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-707721267) you transitioned to C# bots.

We will be attentive to your answer 🙂

**Comment by tomSauret847** [comment]
Thank you for responding @VictorGrycuk  here is the answers to our questions,

1. We are testing with the bots deployed to Azure. We are unable to create a tunnel to test locally.

2. Yes we are using a Azure application gateway with a WAF as outlined in your link

3. These are TypeScript bots (VA and 2 Skills with authentication) we are only using 1 C# skill to get around the synchronous loop [issue](https://github.com/microsoft/botframework-solutions/issues/3626#issuecomment-707721267) and it has no authentication in it.

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, sorry for the delay - we think it **should be possible** to use Enhanced Authentication with an Azure App Gateway having the bots behind it.

A couple of additional **questions**:

1.  Can you connect to the Virtual Assistant with an emulator using the Gateway's public IP? In our testing we confirmed that this is possible
2.  Can you connect and communicate with a skill through the VA **without** enhanced communication?

**Components**:

-   Deployed TypeScript [Virtual Assistant 1.0](https://github.com/microsoft/botframework-solutions/pull/3584)
-   Deployed TypeScript [Skill 1.0](https://github.com/microsoft/botframework-solutions/pull/3585)
-   TypeScript 1.0 [Bot-Solutions](https://github.com/microsoft/botframework-solutions/pull/3583)
-   An [Azure Application Gateway](https://docs.microsoft.com/en-us/azure/web-application-firewall/ag/application-gateway-web-application-firewall-portal) and WAF
-   [Direct Line sample](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web)

**Configuration**:

-   Virtual Assistant using Enhanced Authentication, as explained in [this comment](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-687317382)
-   The Skill connected to the Virtual Assistant configured with the authentication
-   Direct Line as configured in the same comment as above
-   Application Gateway pointing to the deployed bot as explained in [this article](https://docs.microsoft.com/en-us/azure/application-gateway/configure-web-app-portal)

We will be working on replicating this scenario.

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk The answers to your questions are 
Can you connect to the Virtual Assistant with an emulator using the Gateway's public IP? In our testing we confirmed that this is possible

- We do not have tunneling available to we are unable to connect to the VA with a emulator but we have web chat available. The VA will connect to the skill and complete the dialogs.

Can you connect and communicate with a skill through the VA without enhanced communication?

- Yes, if we do not use Enhanced Authentication we can copy and paste the code and the authentication works as expected. We would like the seamless integration that the enhanced authentication provides for our users.

**Comment by VictorGrycuk** [comment]
Hi @tomSauret847, we can **confirm** that Enhanced Authentication **does work** with the bots behind an Azure Application Gateway.

We created the PR [#3694](https://github.com/microsoft/botframework-solutions/pull/3694) that includes a new documentation on how to configure an Azure Application Gateway with a deployed bot.

We did our test by following these **steps**:

1. Follow the steps described in [this comment](https://github.com/microsoft/botframework-solutions/issues/3573#issuecomment-702387057) to deploy bots with Enhanced Authentication
2. Follow the steps detailed in the document [Configuring a Gateway's health probe with a TypeScript bot](https://github.com/microsoft/botframework-solutions/blob/7e4d4791ef7047319561e3efeaa059e6b6a36dc1/docs/_docs/virtual-assistant/handbook/configure-an-azure-application-gateway.md) to connect the deployed Virtual Assistant to the Azure Application Gateway's Backend pool
3. Test the login of the deployed skill using the [Directline sample](https://github.com/microsoft/botframework-solutions/tree/master/samples/csharp/clients/directline.web)
4. The Enhanced Authentication worked and the Skill Sample Dialog continued as expected

As this seems a configuration issue of the Azure Application Gateway, I suggest you to ask in <https://stackoverflow.com/questions/tagged/botframework> for suggestions; and if you are okay, we can **close** this issue 🙂.

**Comment by tomSauret847** [comment]
Thank you @VictorGrycuk We will work through the information that you posted. I did notice already that this [link](https://github.com/microsoft/botframework-solutions/blob/7e4d4791ef7047319561e3efeaa059e6b6a36dc1/docs/_docs/virtual-assistant/handbook/%7B%7Bsite.baseurl%7D%7D/help/known-issues#configuring-a-gateway's-health-probe-with-a-typescript-bot) is broken.

**Comment by VictorGrycuk** [comment]
Thanks @tomSauret847, that link will lead to [Configuring a Gateway's health probe with a TypeScript bot](https://github.com/microsoft/botframework-solutions/blob/7e4d4791ef7047319561e3efeaa059e6b6a36dc1/docs/_docs/help/known-issues.md#configuring-a-gateways-health-probe-with-a-typescript-bot) (as well as [this image](https://github.com/microsoft/botframework-solutions/blob/7e4d4791ef7047319561e3efeaa059e6b6a36dc1/docs/assets/images/help_known_issues_gateway_typescript.jpg)) once the documents are merged into `master` branch, as the documentation uses `{{site.baseurl}}` to generate the URLs.

**Comment by Batta32** [comment]
@peterinnesmsft - we can close the issue due to inactivity. @tomSauret847 if you are still experiencing issues, feel free to reactivate and we can pick this up again, or create a new issue!

Also, if you have any question, please ask in https://stackoverflow.com/questions/tagged/botframework.

The entire authentication process will be solved as soon as the following PRs are merged for TypeScript version 1.0:
- PR #3583: `[TypeScript][Bot-Solutions] Implement changes in Bot-Solutions to 1.0 release`
- PR #3584: `[TypeScript][Virtual Assistant] Implement changes in Virtual Assistant to 1.0 release`
- PR #3585: `[TypeScript][Skill] Implement changes in Skill to 1.0 release`

**Comment by tomSauret847** [comment]
Thank you @Batta32 for your assistance on this. We are having to put this section on hold for the moment so I will close this issue and reopen it if we have issues when we pick it back up again.

**Comment by tomSauret847** [comment]
@Batta32 We have a quick question on this issue. Is the SSO for typescript available yet? We found the issue is our skills are behind a firewall and are unable to receive the token response. When we pass the VA's credentials in it is not forwarding the token to the skill it is just restarting the conversation. If SSO for typescript is not available when do you expect it to be? Thank you for all your help on this issue.

**Comment by Batta32** [comment]
Hi @tomSauret847, SSO is not available for TypeScript currently as it's introduced in 1.0 version. The PRs of TypeScript version 1.0 should be approved and merged in order to release a new version with that feature.

We will let you know as soon as we have any update 😊.
